test_conformance/subgroups/subgroup_common_templates.h

```diff
@@ -567,32 +544,42 @@ template <typename Ty, ShuffleOp operation> struct SHF
                 { // inside the subgroup
                   // shuffle index storage
                     int midx = 4 * ii + 4 * i + 2;
-                    l = (int)m[midx];
+                    l = m[midx];
                     rr = my[ii + i];
+                    cl_uint tr_idx;
+                    bool skip = false;
                     switch (operation)
                     {
                         // shuffle basic - treat l as index
-                        case ShuffleOp::shuffle: tr = mx[ii + l]; break;
-                        // shuffle up - treat l as delta
-                        case ShuffleOp::shuffle_up: tr = mx[ii + i - l]; break;
+                        case ShuffleOp::shuffle: tr_idx = l; break;
+                        // shuffle xor - treat l as mask
+                        case ShuffleOp::shuffle_xor: tr_idx = i ^ l; break;
                         // shuffle up - treat l as delta
-                        case ShuffleOp::shuffle_down:
-                            tr = mx[ii + i + l];
+                        case ShuffleOp::shuffle_up:
+                            if (l >= ns) skip = true;
+                            tr_idx = i - l;
                             break;
-                        // shuffle xor - treat l as mask
-                        case ShuffleOp::shuffle_xor:
-                            tr = mx[ii + (i ^ l)];
+                        // shuffle down - treat l as delta
+                        case ShuffleOp::shuffle_down:
+                            if (l >= ns) skip = true;
+                            tr_idx = i + l;
                             break;
                         default: break;
                     }
 
-                    if (!compare(rr, tr))
+                    if (!skip && tr_idx < n)
```

Comment:
I think this is OK, though perhaps a bit redundant.  In the cases above when `skip` is true, will there ever be a case when the index is valid (`tr_idx < n`)?  If not, debatably the check for `skip` is debatably unnecessary.

Reply:
Yes, there will be such cases. For example, if `i` is 1 and `l` is 0xffffffff, then for _shuffle_up_, `tr_idx` would be 2 and `skip` would be _true_, and for _shuffle_down_, `tr_idx` would be 0 and `skip` would be again be _true_.

The check for `tr_idx < n` will likely be extended in future with a more elaborate test of the "work_item_mask" as used elsewhere, but the use of `skip` would remain unchanged.

Note that where `skip` is set, it is done on the basis of `ns` (the _maximum_ subgroup size), whereas the comparison with `tr_idx` is done against `n` (the _actual_ subgroup size). These relate to separate requirements in the spec, and so in my view, it's cleaner to keep the cases separate here, too.

If we were to move the `tr_idx < n` check into the individual cases, we'd need to merge it into _all four_ cases including _shuffle_ and _shuffle_xor_, with either a somewhat less-than-obvious expression simplification done for _shuffle_up_ and _shuffle_down_, or otherwise with unnecessary duplication of `tr_idx < n` as-is in all four cases.